samples/common/src/misc.rs

```diff
@@ -0,0 +1,220 @@
+///
+/// # Arguments
+/// * `subset` - The provided subset.
+/// * `superset` - The provided superset.
+pub fn is_subset(subset: &[String], superset: &[String]) -> bool {
+    subset.iter().all(|subset_member| {
+        superset.iter().any(|supserset_member| subset_member == supserset_member)
+    })
+}
+
+///
+/// Retry an async function that uses tonic::Status in for its error result.
+///
+/// # Arguments
+/// * `max_retries` - The maximum number of retries.
+/// * `duration_between_attempts` - The duration of time attempts.
+/// * `function` - The function.
+pub async fn retry_async_based_on_status<T, Fut, F: FnMut() -> Fut>(
+    max_retries: i32,
+    duration_between_attempts: Duration,
+    mut function: F,
+) -> Result<T, Status>
+where
+    Fut: Future<Output = Result<T, Status>>,
+{
+    let mut last_status;
+    let mut retries: i32 = 0;
+
+    loop {
+        match function().await {
+            Ok(t) => return Ok(t),
+            Err(status) => {
+                if status.code() == Code::Unavailable || status.code() == Code::Internal {
+                    last_status = status;
+                } else {
+                    return Err(status);
+                }
+            }
+        }
+        if retries < max_retries {
+            debug!("Retrying a call.");
+            sleep(duration_between_attempts).await;
+            retries += 1;
+        } else {
+            break;
+        }
+    }
+
+    Err(last_status)
+}
+
+/// Use Ibeji to discover the endpoint for a digital twin provider that satifies the requirements.
+///
+/// # Arguments
+/// * `invehicle_digitial_twin_servuce_url` - In-vehicle digital twin service URL.
+/// * `entity_id` - The matching entity id.
+/// * `protocol` - The required protocol.
+/// * `operations` - The required operations.
+pub async fn discover_digital_twin_provider_using_ibeji(
+    invehicle_digitial_twin_servuce_url: &str,
+    entity_id: &str,
+    protocol: &str,
+    operations: &[String],
+) -> Result<EndpointInfo, String> {
+    info!("Sending a find_by_id request for entity id {entity_id} to the In-Vehicle Digital Twin Service URL {invehicle_digitial_twin_servuce_url}");
+    let mut client = DigitalTwinClient::connect(invehicle_digitial_twin_servuce_url.to_string())
+        .await
+        .map_err(|error| format!("{error}"))?;
+    let request = tonic::Request::new(FindByIdRequest { id: entity_id.to_string() });
+    let response = client.find_by_id(request).await.map_err(|error| error.to_string())?;
+    let response_inner = response.into_inner();
+    debug!("Received the response for the find_by_id request");
+    info!("response_payload: {:?}", response_inner.entity_access_info);
+
+    let entity_access_info = response_inner.entity_access_info.expect("Did not find the entity");
+
+    let mut matching_endpoint_info_option: Option<EndpointInfo> = None;
+    for endpoint_info in entity_access_info.endpoint_info_list {
```

Comment:
you can use the `find` or `find_map` functions for this: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find

Reply:
Changed.